transport-msg.tex

```diff
@@ -280,45 +297,89 @@ \subsubsection{Bus vs. Transport Messages}
-    Uniquely identifies which message definition applies (e.g., GET_DEVICES,
-    GET_DEVICE_FEATURES, SET_CONFIG). The specific range or enumeration of types is
-    defined in sections \ref{sec:Virtio Transport Options / Virtio Over Messages / Transport Messages}
-    and \ref{sec:Virtio Transport Options / Virtio Over Messages / Bus Messages}.
-  \item \field{dev_num}:
-     Identifies the Device Number the message is targeting or is coming from for
-     Transport Message and must be zero of Bus messages.
-  \item \field{msg_size};
-    Indicates the total length of the message payload including the header.
+  \item \field{msg_id}: Message ID identifying the message definition. Ranges
+    are defined in
+    \ref{sec:Virtio Transport Options / Virtio Over Messages / Transport Messages}
+    and
+    \ref{sec:Virtio Transport Options / Virtio Over Messages / Bus Messages}.
+  \item \field{dev_num}: For Transport Messages, the target device number; for
+    Bus Messages \textbf{MUST} be zero.
+  \item \field{token}: Non-zero for requests (which expect a response); zero
+    \textbf{MUST} be used only for event (one-way) messages. Responses
+    \textbf{MUST} echo the request's \field{token}.
+  \item \field{msg_size}: Total size in bytes of the complete message (header +
+    payload). \textbf{MUST} be \(\ge 8\) and \textbf{MUST NOT} exceed the
+    bus's maximum message size.
+  \item \field{payload}: Operation-specific data. Unused trailing bytes (if any
+    introduced by a bus framing) \textbf{MUST} be zero and \textbf{MUST} be
+    ignored by receivers.
+\end{itemize}
+
+All reserved header bits and any unspecified header values \textbf{MUST} be
+sent as zero and \textbf{MUST} be ignored on receive to preserve forward
+compatibility.
+
+\subsubsection{Message Correlation}
+\label{sec:Virtio Transport Options / Virtio Over Messages / Basic Concepts / Correlation}
+
+Messages are either requests or events. Requests require a response. Events are
+one way and do not have a response. Most requests defined today originate at the
+driver side but bus message requests such as \busref{PING} may originate at
+either driver side or device side.
+
+This section defines how responses are correlated to requests. An implementation
+does not need to support sending multiple requests in parallel but these rules
+allow for that possibility.
+
+The token field in the message header is part of a tuple that is unique during
+a request to response interval.
+\begin{itemize}
+  \item Bus Message tuple: ( 1, 0, \field{token} ).
+  \item Transport Message tuple: ( 0, \field{dev_num}, \field{token} ).
+\end{itemize}
+
+Rules:
+\begin{itemize}
+  \item The request originator assigns a non-zero \field{token} for every
+    request such that the tuple is unique for all inflight requests.
+  \item Event (one-way) messages \textbf{MUST} set \field{token}=0
+  \item A response \textbf{MUST} echo \field{token}, \field{dev_num} (for
+    transport), and \field{msg_id}.
```

Comment:
we might want to remove msg_id here. Correlation should only be based on dev_num and token.
In some cases, the bus might use a different message id for responses.
I have 2 examples in ffa bus:
- event poll which returns an event as response to a poll
- bus error messages which are returned in case of error to a request